### PR TITLE
Throw BadRequest Exception in QuantitySearchValue on malformed input. Fixes #196, #21.

### LINKED ISSUE #196: Search should return 400 instead of 500 when there is FormatException or other exceptions that causes the parsing to fail
If the parsing of the search parameter value fails because of formatting issue or other issue, the search should return 400 instead of 500.

### PR DESCRIPTION
**Pull request by JeffreyTaylor**
## Description
When `parse` fails in QuantitySearchValue, throw a `BadRequestException` exception. This results in an OperationOutcome & 400 status code returned to the user.

## Related issues
#196, #21

## Testing
Updated tests that test this piece of code to now expect a `BadRequestException`.

## Notes
I notice `BadRequestException` is in the `Persistence` namespace. I think it probably should be moved into the `Microsoft.Health.Fhir.Core.Exceptions` namespace.

### CODE REVIEW

File: src/Microsoft.Health.Fhir.Core/Features/Search/SearchValues/QuantitySearchValue.cs [hunk 1]
```diff
@@ -63,14 +64,21 @@ public static QuantitySearchValue Parse(string s)
...
-            string system = parts.Count > 1 ? parts[1] : string.Empty;
-            string code = parts.Count > 2 ? parts[2] : string.Empty;
-
-            return new QuantitySearchValue(
-                system.UnescapeSearchParameterValue(),
-                code.UnescapeSearchParameterValue(),
-                quantity);
+            try
+            {
+                decimal quantity = decimal.Parse(parts[0], NumberStyles.Number, CultureInfo.InvariantCulture);
```
Comment (brandonpollett) [inline src/Microsoft.Health.Fhir.Core/Features/Search/SearchValues/QuantitySearchValue.cs]: If the goal is to catch errors with `Parse`, could we use `TryParse` to catch these issues?

File: src/Microsoft.Health.Fhir.Core/Features/Search/SearchValues/QuantitySearchValue.cs [hunk 2]
```diff
@@ -63,14 +64,21 @@ public static QuantitySearchValue Parse(string s)
...
+                string system = parts.Count > 1 ? parts[1] : string.Empty;
+                string code = parts.Count > 2 ? parts[2] : string.Empty;
+
+                return new QuantitySearchValue(
+                    system.UnescapeSearchParameterValue(),
+                    code.UnescapeSearchParameterValue(),
+                    quantity);
+            }
+            catch
```
Comment (johnstairs) [inline src/Microsoft.Health.Fhir.Core/Features/Search/SearchValues/QuantitySearchValue.cs]: We should avoid catch-all blocks

**Comment by brandonpollett** [comment]
This is related to #196.

File: src/Microsoft.Health.Fhir.Core/Features/Search/SearchValues/QuantitySearchValue.cs [hunk 3]
```diff
@@ -63,14 +64,21 @@ public static QuantitySearchValue Parse(string s)
...
+            {
+                throw new BadRequestException($"Malformed search value \"{parts[0]}\"");
```
Comment (brandonpollett) [inline src/Microsoft.Health.Fhir.Core/Features/Search/SearchValues/QuantitySearchValue.cs]: This text should be in the corresponding resource file.

File: src/Microsoft.Health.Fhir.Core/Features/Search/SearchValues/QuantitySearchValue.cs (hunk 1 above)
Comment (JeffreyTaylor) [reply src/Microsoft.Health.Fhir.Core/Features/Search/SearchValues/QuantitySearchValue.cs]: Good call. I updated this to now use `TryParse`.

File: src/Microsoft.Health.Fhir.Core/Features/Search/SearchValues/QuantitySearchValue.cs (hunk 2 above)
Comment (JeffreyTaylor) [reply src/Microsoft.Health.Fhir.Core/Features/Search/SearchValues/QuantitySearchValue.cs]: Yep, that makes sense. Using `TryParse` solved this issue. This catch-all block has been removed.

File: src/Microsoft.Health.Fhir.Core/Features/Search/SearchValues/QuantitySearchValue.cs (hunk 3 above)
Comment (JeffreyTaylor) [reply src/Microsoft.Health.Fhir.Core/Features/Search/SearchValues/QuantitySearchValue.cs]: Oh - I missed that. This has been fixed. Now using `Core.Resources.MalformedSearchValue`

File: src/Microsoft.Health.Fhir.Core/Resources.resx [hunk 4]
```diff
@@ -326,4 +326,7 @@
   <data name="WeakETagFormatRequired" xml:space="preserve">
     <value>WeakETag must be in the weak ETag format.</value>
   </data>
+    <data name="MalformedSearchValue" xml:space="preserve">
```
Comment (brandonpollett) [inline src/Microsoft.Health.Fhir.Core/Resources.resx]: nit: This file is alphabetized by name to make future diffs easier.

File: src/Microsoft.Health.Fhir.Core/Features/Search/SearchValues/QuantitySearchValue.cs [hunk 5]
```diff
@@ -63,7 +64,12 @@ public static QuantitySearchValue Parse(string s)
                 throw new FormatException(Core.Resources.MoreThanTwoTokenSeparatorSpecified);
             }
 
-            decimal quantity = decimal.Parse(parts[0], NumberStyles.Number, CultureInfo.InvariantCulture);
+            decimal quantity;
+            if (decimal.TryParse(parts[0], NumberStyles.Number, CultureInfo.InvariantCulture, out quantity) == false)
```
Comment (brandonpollett) [inline src/Microsoft.Health.Fhir.Core/Features/Search/SearchValues/QuantitySearchValue.cs]: Elsewhere in the code, we generally use the `!` operator when testing for false on `TryParse`.  
```suggestion
            if (!decimal.TryParse(parts[0], NumberStyles.Number, CultureInfo.InvariantCulture, out quantity))
```

**Pull request merged**